File: android/app/src/main/java/com/macroid/network/SyncClient.kt
package com.macroid.network

import android.util.Log
import com.google.gson.Gson
import com.macroid.util.AppLog
import io.ktor.client.HttpClient
import io.ktor.client.engine.okhttp.OkHttp
import io.ktor.client.request.post
import io.ktor.client.request.setBody
import io.ktor.http.ContentType
import io.ktor.http.contentType
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

class SyncClient(private val deviceFingerprint: String) {

    companion object {
        private const val TAG = "SyncClient"
        private const val MAX_RETRIES = 3
        private const val INITIAL_BACKOFF_MS = 500L
    }

    private val client = HttpClient(OkHttp) {
        engine {
            config {
                connectTimeout(10, java.util.concurrent.TimeUnit.SECONDS)
                readTimeout(30, java.util.concurrent.TimeUnit.SECONDS)
                writeTimeout(30, java.util.concurrent.TimeUnit.SECONDS)
            }
        }
    }

    private val gson = Gson()
    private var peer: DeviceInfo? = null
    @Volatile
    var lastSentText: String = ""
        private set

    fun setPeer(device: DeviceInfo) {
        peer = device
        Log.d(TAG, "Peer set: ${device.alias} at ${device.address}:${device.port}")
        AppLog.add("[SyncClient] Peer set: ${device.alias} at ${device.address}:${device.port}")
    }

    fun sendImage(imageBytes: ByteArray) {
        val device = peer ?: return

        CoroutineScope(Dispatchers.IO).launch {
            val base64 = android.util.Base64.encodeToString(imageBytes, android.util.Base64.NO_WRAP)
            val payload = gson.toJson(
                mapOf(
                    "image" to base64,
                    "timestamp" to System.currentTimeMillis(),
                    "origin" to deviceFingerprint
                )
            )

            AppLog.add("[SyncClient] Sending image (${imageBytes.size} bytes, payload ${payload.length} chars) to ${device.alias}")

            var attempt = 0
            var backoff = INITIAL_BACKOFF_MS
            while (attempt < MAX_RETRIES) {
                try {
                    client.post("http://${device.address}:${device.port}/api/clipboard/image") {
                        contentType(ContentType.Application.Json)
                        setBody(payload)
                    }
                    Log.d(TAG, "Sent image (${imageBytes.size} bytes) to ${device.alias}")
                    AppLog.add("[SyncClient] Sent image successfully to ${device.alias}")
                    return@launch
                } catch (e: Exception) {
                    attempt++
                    if (attempt < MAX_RETRIES) {
                        Log.w(TAG, "Image send failed (attempt $attempt/$MAX_RETRIES), retrying in ${backoff}ms", e)
                        AppLog.add("[SyncClient] Image send failed (attempt $attempt/$MAX_RETRIES): ${e.javaClass.simpleName}: ${e.message}")
                        delay(backoff)
                        backoff *= 2
                    } else {
                        Log.e(TAG, "Image send failed after $MAX_RETRIES attempts", e)
                        AppLog.add("[SyncClient] ERROR: Image send failed after $MAX_RETRIES attempts: ${e.javaClass.simpleName}: ${e.message}")
                    }
                }
            }
        }
    }

    fun sendClipboard(text: String) {
        val device = peer ?: return
        lastSentText = text

        CoroutineScope(Dispatchers.IO).launch {
            val payload = gson.toJson(
                mapOf(
                    "text" to text,
                    "timestamp" to System.currentTimeMillis(),
                    "origin" to deviceFingerprint
                )
            )

            var attempt = 0
            var backoff = INITIAL_BACKOFF_MS

            while (attempt < MAX_RETRIES) {
                try {
                    client.post("http://${device.address}:${device.port}/api/clipboard") {
                        contentType(ContentType.Application.Json)
                        setBody(payload)
                    }
                    Log.d(TAG, "Sent clipboard (${text.length} chars) to ${device.alias}")
                    AppLog.add("[SyncClient] Sent clipboard (${text.length} chars) to ${device.alias}")
                    return@launch
                } catch (e: Exception) {
                    attempt++
                    if (attempt < MAX_RETRIES) {
                        Log.w(TAG, "Send failed (attempt $attempt/$MAX_RETRIES), retrying in ${backoff}ms", e)
                        AppLog.add("[SyncClient] Send failed (attempt $attempt/$MAX_RETRIES): ${e.javaClass.simpleName}: ${e.message}")
                        delay(backoff)
                        backoff *= 2
                    } else {
                        Log.e(TAG, "Send failed after $MAX_RETRIES attempts", e)
                        AppLog.add("[SyncClient] ERROR: Send failed after $MAX_RETRIES attempts: ${e.javaClass.simpleName}: ${e.message}")
                    }
                }
            }
        }
    }
}

File: mac/Macroid/Macroid/SyncClient.swift
import Foundation
import Network
import os.log

private let log = Logger(subsystem: "com.macroid", category: "SyncClient")

class SyncClient {
    private var peer: DeviceInfo
    private let deviceFingerprint: String
    private let maxRetries = 3
    private let queue = DispatchQueue(label: "com.macroid.syncclient")

    init(peer: DeviceInfo, fingerprint: String) {
        self.peer = peer
        self.deviceFingerprint = fingerprint
        AppLog.add("[SyncClient] Initialized for peer: \(peer.alias) at \(peer.address):\(peer.port)")
    }

    func sendClipboard(_ text: String) {
        let payload: [String: Any] = [
            "text": text,
            "timestamp": Int64(Date().timeIntervalSince1970 * 1000),
            "origin": deviceFingerprint
        ]
        guard let body = try? JSONSerialization.data(withJSONObject: payload) else {
            AppLog.add("[SyncClient] Failed to serialize clipboard payload")
            return
        }

        AppLog.add("[SyncClient] Sending clipboard (\(text.count) chars) to \(peer.alias)")
        sendHTTPPost(path: "/api/clipboard", body: body, attempt: 1) { success in
            if success {
                AppLog.add("[SyncClient] Clipboard sent successfully")
            } else {
                AppLog.add("[SyncClient] ERROR: Failed to send clipboard after retries")
            }
        }
    }

    /// Send image notification to peer (pull-based: peer fetches from our server)
    func sendImage(_ imageData: Data, localPort: UInt16) {
        // Build fetch URL that peer can use to download the image
        let localIP = getLocalIPAddress() ?? peer.address
        let fetchURL = "http://\(localIP):\(localPort)/api/image/latest"

        let payload: [String: Any] = [
            "fetch_url": fetchURL,
            "timestamp": Int64(Date().timeIntervalSince1970 * 1000),
            "origin": deviceFingerprint
        ]
        guard let body = try? JSONSerialization.data(withJSONObject: payload) else {
            AppLog.add("[SyncClient] Failed to serialize image notification")
            return
        }

        AppLog.add("[SyncClient] Sending image notification (\(imageData.count) bytes available at \(fetchURL))")
        sendHTTPPost(path: "/api/clipboard/image", body: body, attempt: 1) { success in
            if success {
                AppLog.add("[SyncClient] Image notification sent successfully")
            } else {
                AppLog.add("[SyncClient] ERROR: Failed to send image notification")
            }
        }
    }

    private func getLocalIPAddress() -> String? {
        var address: String?
        var ifaddr: UnsafeMutablePointer<ifaddrs>?
        guard getifaddrs(&ifaddr) == 0, let firstAddr = ifaddr else { return nil }
        defer { freeifaddrs(ifaddr) }
        for ptr in sequence(first: firstAddr, next: { $0.pointee.ifa_next }) {
            let sa = ptr.pointee.ifa_addr.pointee
            guard sa.sa_family == UInt8(AF_INET) else { continue }
            let name = String(cString: ptr.pointee.ifa_name)
            guard name == "en0" || name == "en1" else { continue }
            var hostname = [CChar](repeating: 0, count: Int(NI_MAXHOST))
            getnameinfo(ptr.pointee.ifa_addr, socklen_t(sa.sa_len), &hostname, socklen_t(hostname.count), nil, 0, NI_NUMERICHOST)
            address = String(cString: hostname)
            break
        }
        return address
    }

    private func sendHTTPPost(path: String, body: Data, attempt: Int, completion: @escaping (Bool) -> Void) {
        guard let port = NWEndpoint.Port(rawValue: UInt16(peer.port)) else {
            AppLog.add("[SyncClient] Invalid port: \(peer.port)")
            completion(false)
            return
        }

        let host = NWEndpoint.Host(peer.address)
        let connection = NWConnection(host: host, port: port, using: .tcp)
        var completed = false

        let httpRequest = "POST \(path) HTTP/1.1\r\nHost: \(peer.address):\(peer.port)\r\nContent-Type: application/json\r\nContent-Length: \(body.count)\r\nConnection: close\r\n\r\n"
        var requestData = httpRequest.data(using: .utf8)!
        requestData.append(body)

        connection.stateUpdateHandler = { [weak self] state in
            guard let self = self, !completed else { return }
            switch state {
            case .ready:
                AppLog.add("[SyncClient] TCP connected to \(self.peer.address):\(self.peer.port) for \(path)")
                connection.send(content: requestData, completion: .contentProcessed { error in
                    if let error = error {
                        AppLog.add("[SyncClient] Send error: \(error.localizedDescription)")
                        completed = true
                        connection.cancel()
                        self.retryIfNeeded(path: path, body: body, attempt: attempt, completion: completion)
                        return
                    }
                    // Read response
                    connection.receive(minimumIncompleteLength: 1, maximumLength: 65536) { data, _, _, error in
                        completed = true
                        connection.cancel()
                        if let data = data, let response = String(data: data, encoding: .utf8) {
                            let statusLine = response.components(separatedBy: "\r\n").first ?? ""
                            AppLog.add("[SyncClient] Response: \(statusLine)")
                            completion(statusLine.contains("200"))
                        } else {
                            AppLog.add("[SyncClient] No response data: \(error?.localizedDescription ?? "unknown")")
                            self.retryIfNeeded(path: path, body: body, attempt: attempt, completion: completion)
                        }
                    }
                })
            case .failed(let error):
                AppLog.add("[SyncClient] TCP connection failed: \(error.localizedDescription)")
                completed = true
                connection.cancel()
                self.retryIfNeeded(path: path, body: body, attempt: attempt, completion: completion)
            case .cancelled:
                if !completed {
                    completed = true
                    completion(false)
                }
            default:
                break
            }
        }

        connection.start(queue: queue)

        // Timeout - longer for image uploads
        let timeout: Double = path.contains("image") ? 30 : 5
        queue.asyncAfter(deadline: .now() + timeout) {
            if !completed {
                completed = true
                AppLog.add("[SyncClient] Timeout sending to \(path)")
                connection.cancel()
                self.retryIfNeeded(path: path, body: body, attempt: attempt, completion: completion)
            }
        }
    }

    private func retryIfNeeded(path: String, body: Data, attempt: Int, completion: @escaping (Bool) -> Void) {
        if attempt < maxRetries {
            let backoff = Double(1 << (attempt - 1)) * 0.5
            AppLog.add("[SyncClient] Retrying \(path) in \(backoff)s (attempt \(attempt)/\(maxRetries))")
            queue.asyncAfter(deadline: .now() + backoff) {
                self.sendHTTPPost(path: path, body: body, attempt: attempt + 1, completion: completion)
            }
        } else {
            AppLog.add("[SyncClient] ERROR: All \(maxRetries) retries exhausted for \(path)")
            completion(false)
        }
    }
}
